src/level/level_loader.cpp

```diff
@@ -0,0 +1,1201 @@
+{
+    std::unique_ptr<CLevelLoader> levelLoader = MakeUnique<CLevelLoader>(solutionEnabled);
+
+    levelLoader->ResetScene(mode);
+
+    levelLoader->BuildScene(fixScene, mode);
+}
+
+void CLevelLoader::ResetScene(LevelLoadMode mode)
+{
+    m_main->m_base = nullptr;
+
+    if (mode != LevelLoadMode::Reset)
+    {
+        m_main->m_build = 0;
+        m_main->m_researchDone.clear();  // no research done
+        m_main->m_researchDone[0] = 0;
+        m_main->m_researchEnable = 0;
+
+        m_main->ResetUnit();
+
+        m_main->FlushDisplayInfo();
+        m_main->m_terrain->FlushMaterials();
+        m_main->m_audioTrack = "";
+        m_main->m_audioRepeat = true;
+        m_main->m_satcomTrack  = "";
+        m_main->m_satcomRepeat = true;
+        m_main->m_editorTrack  = "";
+        m_main->m_editorRepeat = true;
+        m_main->m_displayText->SetDelay(1.0f);
+        m_main->m_displayText->SetEnable(true);
+        m_main->m_immediatSatCom = false;
+        m_main->m_lockedSatCom = false;
+        m_main->m_endingWin = "";
+        m_main->m_endingLost = "";
+        m_main->m_audioChange.clear();
+        m_main->m_endTake.clear();
+        m_main->m_endTakeImmediat = false;
+        m_main->m_endTakeResearch = 0;
+        m_main->m_endTakeWinDelay = 2.0f;
+        m_main->m_endTakeLostDelay = 2.0f;
+        m_main->m_globalMagnifyDamage = 1.0f;
+        m_main->m_obligatoryTokens.clear();
+        m_main->m_mapShow = true;
+        m_main->m_mapImage = false;
+        m_main->m_mapFilename[0] = 0;
+
+        m_main->m_controller = nullptr;
+
+        m_main->ResetColors();
+
+        m_main->m_engine->SetAmbientColor(Gfx::Color(0.5f, 0.5f, 0.5f, 0.5f), 0);
+        m_main->m_engine->SetAmbientColor(Gfx::Color(0.5f, 0.5f, 0.5f, 0.5f), 1);
+        m_main->m_engine->SetFogColor(Gfx::Color(1.0f, 1.0f, 1.0f, 1.0f), 0);
+        m_main->m_engine->SetFogColor(Gfx::Color(1.0f, 1.0f, 1.0f, 1.0f), 1);
+        m_main->m_engine->SetDeepView(1000.0f, 0);
+        m_main->m_engine->SetDeepView(1000.0f, 1);
+        m_main->m_engine->SetFogStart(0.75f, 0);
+        m_main->m_engine->SetFogStart(0.75f, 1);
+        m_main->m_engine->SetSecondTexture("");
+        m_main->m_engine->SetForegroundName("");
+
+        GetResource(RES_TEXT, RT_SCRIPT_NEW, m_main->m_scriptName);
+        m_main->m_scriptFile = "";
+
+        m_main->m_missionType   = MISSION_NORMAL;
+        m_main->m_codeBattleInit = false;
+        m_main->m_codeBattleStarted = false;
+
+        m_main->m_teamNames.clear();
+
+        m_main->m_missionResult = ERR_MISSION_NOTERM;
+        m_main->m_missionResultFromScript = false;
+    }
+
+    // NOTE: Reset timer always, even when only resetting object positions
+    m_main->m_missionTimerEnabled = false;
+    m_main->m_missionTimerStarted = false;
+    m_main->m_missionTimer = 0.0f;
+}
+
+void CLevelLoader::BuildScene(bool fixScene, LevelLoadMode mode)
+{
+    m_main->m_fixScene = fixScene;
```

Comment:
Setting fixScene variable should not be part of loading either, this should be set in CRobotMain itself before calling the loader.